lib/elixir/lib/kernel.ex

```diff
@@ -2642,6 +2642,126 @@ defmodule Kernel do
     end
   end
 
+  @doc """
+  Evaluates the expression corresponding to the first clause that
+  evaluates to a truthy value.
+
+      cond do
+        hd([1, 2, 3]) ->
+          "1 is considered as true"
+      end
+      #=> "1 is considered as true"
+
+  Raises an error if all conditions evaluate to `nil` or `false`.
+  For this reason, it may be necessary to add a final always-truthy condition
+  (anything non-`false` and non-`nil`), which will always match.
+
+  ## Examples
+
+      cond do
+        1 + 1 == 1 ->
+          "This will never match"
+        2 * 2 != 4 ->
+          "Nor this"
+        true ->
+          "This will"
+      end
+      #=> "This will"
+
+  """
+
+  defmacro cond(clauses) do
+    build_cond(clauses, __CALLER__)
+  end
+
+  defp build_cond([do: do_clause], caller) do
+    %{line: line} = caller
+
+    [{:->, meta, [[condition], body]} | tail] =
+      reversed_clauses = reverse_cond_clauses(do_clause, [])
+
+    {next_clauses, next_acc} =
+      case condition do
+        {:_, _, context} when is_atom(context) ->
+          raise ArgumentError,
+                "invalid use of _ inside \"cond\". If you want the last clause " <>
+                  "to always match, you probably meant to use: true ->"
+
+        value when is_atom(value) and value != false and value != nil ->
+          {tail, body}
+
+        _ ->
+          error = quote(do: raise(CondClauseError))
+          {reversed_clauses, error}
+      end
+
+    case_clause = build_cond_clauses(next_clauses, next_acc, meta)
+    replace_case_line(case_clause, line)
```

Comment:
Do we need to do this? It is fine to me if we always use the meta of the `->` and never the meta of the `cond` call itself.

Reply:
[There was an explicit test for that](https://github.com/elixir-lang/elixir/blob/master/lib/elixir/test/erlang/control_test.erl#L186), so I was afraid to change. I don't think there are big implications of that, maybe a better error message when there are problems in the first condition.